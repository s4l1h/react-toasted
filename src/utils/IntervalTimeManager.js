const IntervalTimeManager = params => ({
  id: false,
  times: {},
  estimated: null,
  remaning: null,
  totalTime: params.totalTime || 5000,
  stepTime: params.stepTime || 50,
  callbackFunctions: params.callbackFunctions || {},
  callback() {
    this.times["callback"] = this.getTime();

    this.remaning = this.remaning - this.stepTime;
    this.estimated = this.estimated + this.stepTime;

    this.callCallbackFN("callback");

    if (this.remaning <= 0) {
      return this.finish();
    }
  },
  getTime() {
    return new Date().getTime();
  },
  getPercent() {
    return Math.floor((this.remaning / this.totalTime) * 100);
  },
  start() {
    this.times["started"] = this.getTime();
    this.callCallbackFN("before:start");
    this.remaning = this.totalTime;
    this._setupInterval();
    this.callCallbackFN("after:start");
  },
  finish() {
    this.times["finished"] = this.getTime();
    this.callCallbackFN("before:finish");
    this._clearInterval(this.id);
    this.callCallbackFN("after:finish");
  },
  stop() {
    this.times["stoped"] = this.getTime();
    // People can stop manualy
    this.callCallbackFN("before:stop");
    this._clearInterval(this.id);
    this.callCallbackFN("after:stop");
  },
  pause() {
    this.times["paused"] = this.getTime();
    this.callCallbackFN("before:pause");
    this._clearInterval(this.id);
    this.callCallbackFN("after:pause");
  },
  resume() {
    this.times["resumed"] = this.getTime();
    this.callCallbackFN("before:resume");
    this._setupInterval();
    this.callCallbackFN("after:resume");
  },
  callCallbackFN(type) {
    // console.log(this.callbackFunctions, type);
    if (typeof this.callbackFunctions[type] === "function") {
      this.callbackFunctions[type]();
    }
  },
  _clearInterval() {
    clearInterval(this.id);
  },
  _setupInterval() {
    this.id = setInterval(() => {
      this.callback();
    }, this.stepTime);
  }
});
export default IntervalTimeManager;
