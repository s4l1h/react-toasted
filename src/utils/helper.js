export const generateList = positions => {
  const list = {};
  for (let i = 0; i <= positions.length - 1; i += 1) {
    list[positions[i]] = [];
  }
  return list;
};
// Thanks Dan
// hhttps://twitter.com/dan_abramov/status/824308413559668744
export const addToast = newData => {
  return state => {
    const { list, savedNames } = state;
    if (typeof newData.name !== "undefined") {
      savedNames[newData.name] = newData.index;
    }

    list[newData.position] = newData.newestOnTop
      ? [newData, ...list[newData.position]]
      : [...list[newData.position], newData];

    return { list, savedNames };
  };
};
export const removeToast = item => {
  return state => {
    const list = { ...state.list };
    const savedNames = { ...state.savedNames };
    list[item.position] = list[item.position].filter(toast => {
      return toast.index !== item.index;
    });

    const indexList = Object.values(savedNames);
    const nameList = Object.keys(savedNames);
    if (indexList.includes(item.index)) {
      for (let i = 0; i <= nameList.length - 1; i += 1) {
        const name = nameList[i];
        if (savedNames[name] === item.index) {
          delete savedNames[name];
        }
      }
    }
    // delete list[item.position][item.index];
    return { list, savedNames };
  };
};
export const findToastByIndex = (list, index) => {
  let item = [];
  const positions = Object.keys(list);
  for (let i = 0; i <= positions.length - 1; i += 1) {
    const position = positions[i];
    item = list[position].filter(toast => toast.index === index);
    if (item.length !== 0) {
      break;
    }
  }
  return item;
};
