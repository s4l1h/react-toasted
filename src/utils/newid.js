let lastId = 0;

export default () => {
  lastId++;
  return lastId;
};
