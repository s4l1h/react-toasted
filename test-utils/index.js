import React, { useContext } from "react";
import { render } from "@testing-library/react";
import ToastedContext, { Provider, Consumer, withToasted } from "../src/index";

export const SampleConsumerUseContext = () => {
  // Acces toast from context
  const toasted = useContext(ToastedContext);
  return <BasicComponent toasted={toasted} />;
};
export class SampleConsumerContextType extends React.Component {
  // use toasted contextType
  static contextType = ToastedContext;

  render() {
    // get toasted from context
    const toasted = this.context;
    return <BasicComponent toasted={toasted} />;
  }
}
export const SampleConsumerContext = () => {
  return (
    <Consumer>
      {toasted => {
        return <BasicComponent toasted={toasted} />;
      }}
    </Consumer>
  );
};
export const BasicComponent = ({ toasted: { add, remove, removeAll } }) => {
  return (
    <React.Fragment>
      <button
        onClick={() => {
          add({
            name: "mytoast",
            title: "First Toast",
            msg: "First Toast Message"
          });
          add({
            title: "Second Toast",
            msg: "Second Toast Message"
          });
        }}
      >
        Create Toasts
      </button>
      <button
        onClick={() => {
          remove("mytoast");
        }}
      >
        Remove Toast
      </button>
      <button
        onClick={() => {
          removeAll();
        }}
      >
        Remove All
      </button>
    </React.Fragment>
  );
};

export const WithToastedComponent = withToasted(BasicComponent);

export const ProviderWrapper = ({ children }) => {
  return (
    <Provider
      {...{
        defaultClassNames: ["animated", "zoomInUp"], // need animated.css
        defaultPosition: "toast-top-right",
        defaultNewestOnTop: false,
        defaultType: "error",
        defaultRemoveOnClick: true,
        defaultTimeout: 5000,
        defaultProgressBar: true,
        defaultProgressBarValue: 0,
        defaultPreventDuplicates: false,
        defaultStyle: {}
        // defaultPositions: positions
      }}
    >
      {children}
    </Provider>
  );
};

const renderWithProvider = (ui, options) =>
  render(ui, { wrapper: ProviderWrapper, ...options });

// override render method
export { renderWithProvider };
