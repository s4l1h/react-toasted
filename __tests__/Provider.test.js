// __tests__/hidden-message.js
// these imports are something you'd normally configure Jest to import for you
// automatically. Learn more in the setup docs: https://testing-library.com/docs/react-testing-library/setup#cleanup
// import "@testing-library/react/cleanup-after-each";
import "@testing-library/jest-dom/extend-expect";

import { renderWithProvider } from "../src/utils/test-utils";

describe("Toasted Provider Snapshot", () => {
  test("it has to match snapshot", () => {
    const { container } = renderWithProvider();
    expect(container).toMatchSnapshot();
  });
});
