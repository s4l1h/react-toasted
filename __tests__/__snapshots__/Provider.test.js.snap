// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toasted Provider Snapshot it has to match snapshot 1`] = `
<div>
  <div
    class="toast-container toast-top-right"
  />
  <div
    class="toast-container toast-bottom-right"
  />
  <div
    class="toast-container toast-bottom-left"
  />
  <div
    class="toast-container toast-top-left"
  />
  <div
    class="toast-container toast-top-full-width"
  />
  <div
    class="toast-container toast-bottom-full-width"
  />
  <div
    class="toast-container toast-top-center"
  />
  <div
    class="toast-container toast-bottom-center"
  />
</div>
`;
